// fd/model/properties.js
exports.initial = {
  alignmentBaseline: 'auto',
  baselineShift: 'baseline',
  clip: 'auto',
  clipPath: 'none',
  clipRule: 'nonzero',
  color: 'black',
  colorInterpolation: 'sRGB',
  colorInterpolationFilters: 'linearRGB',
  colorProfile: 'auto',
  cursor: 'auto',
  direction: 'ltr',
  display: 'inline',
  dominantBaseline: 'auto',
  enableBackground: 'accumulate',
  fill: 'black',
  fillOpacity: 1,
  fillRule: 'nonzero',
  filter: 'none',
  floodColor: 'black',
  floodOpacity: 1,
  fontFamily: 'serif',
  fontSize: 'medium',
  fontSizeAdjust: 'none',
  fontStretch: 'normal',
  fontStyle: 'normal',
  fontVariant: 'normal',
  fontWeight: 'normal',
  glyphOrientationHorizontal: '0deg',
  glyphOrientationVertical: 'auto',
  imageRendering: 'auto',
  kerning: 'auto',
  letterSpacing: 'normal',
  lightingColor: 'white',
  markerEnd: 'none',
  markerMid: 'none',
  markerStart: 'none',
  mask: 'none',
  opacity: 1,
  overflow: 'hidden',
  pointerEvents: 'visiblePainted',
  shapeRendering: 'auto',
  stopColor: 'black',
  stopOpacity: 1,
  stroke: 'none',
  strokeDasharray: 'none',
  strokeDashoffset: 0,
  strokeLinecap: 'butt',
  strokeLinejoin: 'miter',
  strokeMiterlimit: 4,
  strokeOpacity: 1,
  strokeWidth: 1,
  textAnchor: 'start',
  textDecoration: 'none'
}

exports.inherited = [
  'clipRule',
  'color',
  'colorInterpolation',
  'colorInterpolationFilters',
  'colorProfile',
  'cursor',
  'direction',
  'fill',
  'fillOpacity',
  'fillRule',
  'fontFamily',
  'fontSize',
  'fontSizeAdjust',
  'fontStretch',
  'fontStyle',
  'fontVariant',
  'fontWeight',
  'glyphOrientationHorizontal',
  'glyphOrientationVertical',
  'imageRendering',
  'kerning',
  'letterSpacing',
  'markerEnd',
  'markerMid',
  'markerStart',
  'pointerEvents',
  'stroke',
  'strokeDasharray',
  'strokeDashoffset',
  'strokeLinecap',
  'strokeLinejoin',
  'strokeMiterlimit',
  'strokeOpacity',
  'strokeWidth',
  'textAnchor',
  'textDecoration',
  'textRendering',
  'visibility',
  'wordSpacing',
  'writingMode'
]
