// fd/ui/command-overlay.js
var h = require('fd/html')
var Native = require('native')
var commands = require('fd/commands')
var commandList = require('fd/command-list')
var keyBindings = require('fd/key-bindings')
var displayBinding = require('fd/display-binding')

function CommandOverlay(editor) {
  this.editor = editor
  this._visible = false

  this.el = h('.command-overlay', [
    this.input = h('input', {placeholder: 'Type a command…'}),
    this.list = h('ul.command-list')
  ])

  this.input.addEventListener('blur', this.onBlur.bind(this))
  this.input.addEventListener('keydown', this.onKeyDown.bind(this))
  this.input.addEventListener('input', this.update.bind(this))
  this.list.addEventListener('click', this.onListClick.bind(this))
  this.list.addEventListener('mousedown', this.onMouseDown.bind(this))
  commands.add('showCommands', {
    title: 'Show Commands',
    fn: this.show
  }, this)
}

CommandOverlay.prototype.onBlur = function(e) {
  if (this.ignoreBlur) {
    this.ignoreBlur = false
    this.input.focus()
    return
  }
  this.hide()
}

CommandOverlay.prototype.onMouseDown = function(e) {
  this.ignoreBlur = true
}

CommandOverlay.prototype.onKeyDown = function(e) {
  if (e.keyIdentifier === 'U+001B') {
    this.hide()
  } else if (e.keyIdentifier === 'Enter') {
    this.accept()
  } else if (e.keyIdentifier === 'U+0020' && this.input.selectionStart === 0) {
  } else {
    return
  }
  e.preventDefault()
}

CommandOverlay.prototype.onListClick = function(e) {
  var t = e.target
  while (t.tagName !== 'LI') {
    if (t.tagName === 'UL') return
    t = t.parentNode
    if (!t) return
  }
  if (t.dataset.name) {
    if (t.classList.contains('enabled')) {
      Native.hooks.runCommand(t.dataset.name)
      this.hide()
    }
    return
  }
  t.classList.toggle('expanded')
  this.input.focus()
}

CommandOverlay.prototype.accept = function() {
  this.hide()
}

CommandOverlay.prototype.show = function() {
  this.visible = true
}
CommandOverlay.prototype.hide = function() {
  this.visible = false
}

Object.defineProperty(CommandOverlay.prototype, 'visible', {
  get: function() {return this._visible},
  set: function(value) {
    this.el.classList.toggle('show', this._visible = value)
    if (value) {
      this.update()
      this.input.select()
    }
  }
})

var bindingFor = function(map) {
  for (var binding in keyBindings) {
    var name = keyBindings[binding]
    if (!map[name]) map[name] = binding
  }
  return map
}(Object.create(null))

CommandOverlay.prototype.update = function() {
  var q = this.input.value
  while (this.list.firstChild) this.list.removeChild(this.list.lastChild)

  if (!q) return
  var list = commandList.commands.map(function(it) {
    return {score: this.scoreItem(it, q), item: it}
  }, this).filter(function(it) {
    return it.score >= 0
  })
  list.sort(function(a, b) {
    return a.score - b.score
  })
  list.forEach(this.addScoredItem, this)
}

CommandOverlay.prototype.scoreItem = function(it, q) {
  var c = commands.get(it.name)
  var i = c.title.toLowerCase().indexOf(q.toLowerCase())
  var j = (it.path.join(' ') + ' ' + c.title).toLowerCase().indexOf(q.toLowerCase())
  if (j !== -1) j = (j + 1) * 10
  return i === -1 ? j : j === -1 ? i : Math.min(i, j)
}

CommandOverlay.prototype.addScoredItem = function(o) {
  var it = o.item
  if (!it.name) return // TODO
  var binding = bindingFor[it.name]
  var c = commands.get(it.name)
  this.list.appendChild(h('li.command' + (c.enabled ? ' enabled' : ''), {
    dataset: {name: it.name}
  }, [
    h('.title', [
      h('span.path', it.path.map(function(name) {
        return h('span.path-item', [name])
      })),
      c.title,
      h('.binding', [binding ? displayBinding(binding) : '']),
    ]),
  ]))
}

module.exports = CommandOverlay
