// fd/ui/editor.js
var h = require('fd/html')
var Vec2 = require('fd/vec2')
var emitter = require('fd/emitter')
var commands = require('fd/commands')
var M = require('fd/model')

var Panel = require('fd/ui/panel')
var Canvas = require('fd/ui/canvas')
var Inspector = require('fd/ui/inspector')

function Editor() {
  this.layers = new Panel(215, h('.layers panel', [
    h('.panel-header'),
    h('.panel-content'),
    h('.panel-footer')
  ]))

  this.canvas = new Canvas(this)
  this.inspector = new Inspector(this)

  this._selection = null

  this.el = h('.editor', [
    this.layers.el,
    this.canvas.el,
    this.inspector.el
  ])

  commands.on('selectParent', this.selectParent, this)
}

emitter(Editor.prototype)
emitter.property(Editor.prototype, 'selection')

Editor.prototype.selectParent = function() {
  var s = this.selection
  if (!s) return true
  this.selection = s.parent && s.parent.parent ? s.parent : null
}

Editor.prototype.start = function() {
  document.body.appendChild(this.el)
  addEventListener('resize', this.onResize.bind(this))
  this.layers.on('resize', this.onResize, this)
  this.onResize()
  this.canvas.center = new Vec2(30, -20)
  this.canvas.scale = 1.5
  this.canvas.onResize()
  return this
}

Editor.prototype.onResize = function() {
  this.canvas.onResize()
}

Editor.prototype.getData = function(type) {
  throw new Error('Unimplemented')
}

Editor.prototype.loadData = function(type, data) {
  return // TODO implement M.load
  if (type !== 'public.svg-image') throw new Error('Unimplemented')
  var document = new DOMParser().parseFromString(data, 'image/svg+xml')
  this.canvas.setDocument(M.load(document))
}

module.exports = Editor
