// fd/ui/tree/index.js
var h = require('fd/html')
var Node = require('./node')

function Tree(editor) {
  this.editor = editor
  this.nodes = new WeakMap

  this.el = h('.tree panel', [
    h('.panel-header'),
    this.content = h('.panel-content'),
    h('.panel-footer'),
  ])

  editor.on('documentChange', this.onDocumentChange, this)
  this.onDocumentChange()
}

Tree.prototype.onDocumentChange = function() {
  if (this.document) this.document.unlisten('change', this.update, this)
  this.document = this.editor.document
  this.document.on('change', this.update, this)
  this.reload()
}

Tree.prototype.update = function(changes) {
  for (var i = changes.length; i--;) {
    var c = changes[i]
    var n = this.nodes.get(c.parent)
    if (!n) continue
    n.update(c)
  }
}

Tree.prototype.reload = function() {
  if (this.children) {
    for (var i = this.children.length; i--;) {
      this.children[i].detach()
    }
  }
  this.children = []
  for (var children = this.document.children, i = 0, l = children.length; i < l; i++) {
    var n = new Node(this, children[i])
    this.content.appendChild(n.el)
    this.children.push(n)
  }
}

module.exports = Tree
