// fd/ui/tree/node.js
var h = require('fd/html')

function Node(tree, model) {
  this.tree = tree
  this.model = model
  this.open = false
  this.loaded = false
  this.children = null

  this.el = h('.node', [
    this.item = h('.item', [
      this.disclosureButton = model.isGroup ? h('button.disclosure') : h('.disclosure placeholder'),
      this.icon = h('.icon'),
      this.title = h('.title'),
      this.lock = h('button.lock'),
      this.hide = h('button.hide'),
    ]),
    this.container = h('.children'),
  ])

  this.updateIcon()
  this.updateTitle()
  this.updateHidden()

  model.on('nameChange', this.updateName)
  model.on('hiddenChange', this.updateHidden)
  this.disclosureButton.addEventListener('click', this.toggle.bind(this))

  tree.nodes.set(model, this)
  if (this.model.name === 'Group 20') this.toggle()
}

Node.prototype.updateIcon = function() {}
Node.prototype.updateTitle = function() {
  this.title.textContent = this.model.name
}
Node.prototype.updateHidden = function() {}

Node.prototype.update = function(c) {
  if (!this.loaded) return
  switch (c.kind) {
    case 'append':
      this.container.appendChild(this.makeNode(c.child).el)
      return
    case 'insert':
      this.container.insertBefore(this.makeNode(c.child).el, this.container.childNodes[c.index])
      return
    case 'replace':
      this.container.replaceChild(this.makeNode(c.child).el, this.tree.nodes.get(c.old).detach().el)
      return
    case 'replaceAll':
      var f = document.createDocumentFragment()
      c.children.forEach(function(child) {
        f.appendChild(this.makeNode(child).el)
      }, thsi)
      this.container.replaceChild(f, this.tree.nodes.get(c.old).detach().el)
      return
    case 'remove':
      this.container.removeChild(this.tree.nodes.get(c.child).detach().el)
      return
  }
}

Node.prototype.detach = function() {
  this.model.unlisten('nameChange', this.updateName)
  this.model.unlisten('hiddenChange', this.updateHidden)
  this.tree.nodes.delete(this)
  return this
}

Node.prototype.toggle = function() {
  this.item.classList.toggle('open', this.open = !this.open)
  if (this.loaded || !this.open) {
    this.container.style.display = this.open ? 'block' : 'none'
    return
  }
  this.loaded = true
  this.children = this.model.children.map(function(child) {
    var n = this.makeNode(child)
    this.container.appendChild(n.el)
    return n
  }, this)
}

Node.prototype.makeNode = function(child) {
  return new Node(this.tree, child)
}

module.exports = Node
