// fd/ui/tree/node.js
var h = require('fd/html')

function Node(tree, model) {
  this.model = model
  this.open = false
  this.loaded = false
  this.nodes = null

  this.el = h('.node', [
    this.item = h('.item', [
      this.disclosureButton = model.isGroup ? h('button.disclosure') : h('.disclosure placeholder'),
      this.icon = h('.icon'),
      this.title = h('.title'),
      this.lock = h('button.lock'),
      this.hide = h('button.hide'),
    ]),
    this.children = h('.children'),
  ])

  this.updateIcon()
  this.updateTitle()
  this.updateHidden()

  model.on('nameChange', this.updateName)
  model.on('hiddenChange', this.updateHidden)
  this.disclosureButton.addEventListener('click', this.toggle.bind(this))
  this.toggle()
}

Node.prototype.updateIcon = function() {}
Node.prototype.updateTitle = function() {
  this.title.textContent = this.model.name
}
Node.prototype.updateHidden = function() {}

Node.prototype.toggle = function() {
  this.item.classList.toggle('open', this.open = !this.open)
  if (this.loaded || !this.open) {
    this.children.style.display = this.open ? 'block' : 'none'
    return
  }
  this.loaded = true
  this.nodes = this.model.children.map(function(child) {
    var n = new Node(this.tree, child)
    this.children.appendChild(n.el)
    return n
  }, this)
}

module.exports = Node
